onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -0,0 +1,721 @@
+                                                  const NchwcArgument::Shape& input_shape,
+                                                  NchwcArgument::Shape& output_shape,
+                                                  const ONNX_NAMESPACE::TensorProto* filter_shape) {
+  // Skip the leading batch and channel counts.
+  const int kernel_size = kNchwcDims - 2;
+
+  // Maintain the batch count dimension from the NCHWc input.
+  output_shape.dims_[0] = input_shape.dims_[0];
+
+  const ONNX_NAMESPACE::AttributeProto* pads_attr = GetAttribute(node, "pads");
+  const ONNX_NAMESPACE::AttributeProto* strides_attr = GetAttribute(node, "strides");
+  const ONNX_NAMESPACE::AttributeProto* dilations_attr = GetAttribute(node, "dilations");
+
+  if ((pads_attr != nullptr && pads_attr->ints_size() != kernel_size * 2) ||
+      (strides_attr != nullptr && strides_attr->ints_size() != kernel_size) ||
+      (dilations_attr != nullptr && dilations_attr->ints_size() != kernel_size)) {
+    return;
+  }
+
+  // Require the kernel_shape attribute for pooling operators. Convolution
+  // uses the weight tensor shape to derive the kernel shape.
+  const ONNX_NAMESPACE::AttributeProto* kernel_shape_attr = nullptr;
+  if (filter_shape == nullptr) {
+    kernel_shape_attr = GetAttribute(node, "kernel_shape");
+    if (kernel_shape_attr == nullptr || kernel_shape_attr->ints_size() != kernel_size) {
+      return;
+    }
+  }
+
+  auto* auto_pad_attr = GetAttribute(node, "auto_pad");
+  bool auto_pad_same_shape = false;
+  if (auto_pad_attr != nullptr && auto_pad_attr->has_s()) {
+    auto& auto_pad = auto_pad_attr->s();
+    if (auto_pad != "NOTSET") {
+      if (auto_pad == "SAME_UPPER" || auto_pad == "SAME_LOWER") {
+        auto_pad_same_shape = true;
+      } else if (auto_pad != "VALID") {
+        return;
+      }
+      pads_attr = nullptr;
+    }
+  }
+
+  for (int i = 0; i < kernel_size; i++) {
+    if ((strides_attr != nullptr && strides_attr->ints(i) != 1) ||
+        (dilations_attr != nullptr && dilations_attr->ints(i) != 1)) {
+      continue;
+    }
+
+    int64_t padding = 0;
+    if (pads_attr != nullptr) {
+      padding = pads_attr->ints(i) + pads_attr->ints(i + kernel_size);
+    }
+
+    int64_t kernel;
+    if (kernel_shape_attr != nullptr) {
+      kernel = kernel_shape_attr->ints(i);
+    } else {
+      kernel = filter_shape->dims(2 + i);
+    }
+
+    if (padding + 1 == kernel || auto_pad_same_shape) {
+      output_shape.dims_[2 + i] = input_shape.dims_[2 + i];
+    }
+  }
+}
+
+void NchwcTransformerImpl::TransformConv(Node& node) {
+  auto& input_defs = node.MutableInputDefs();
+  auto& output_defs = node.MutableOutputDefs();
+
+  // Require that the weights tensor be static.
+  const ONNX_NAMESPACE::TensorProto* conv_W_tensor_proto = nullptr;
+  if (!graph_utils::NodeArgIsConstant(graph_, *input_defs[1]) ||
+      !graph_.GetInitializedTensor(input_defs[1]->Name(), conv_W_tensor_proto) ||
```

Comment:
NodeArgIsConstant was just updated to handle a subgraph consuming an initializer from a parent graph. Does GetInitializedTensor also need to be able to retrieve an initializer from a parent graph if necessary?

Reply:
I'm doing the same logic as the other optimizers (I'm looking at conv_bn_fusion now). Do you have a recommended pattern to follow here? I can also add a test to nchwc_optimizer_test.cc that optimizes a subgraph. If I add a test based on if_test.cc, does that hit the needed case?

Reply:
The other optimizers probably need updating too. May be more useful to have a GetConstantInitializer in graph_utils instead of separate NodeArgIsConstant and GetInitializedTensor calls. I'll put together a quick PR for that tonight. You can either pickup some or all of the the changes from there or wait for it (Konstantinos may want to take a look and he's out until Monday).

A test based on if_test.cc would work. You can just use the same subgraph for both the then_branch and else_branch. Key thing is to call AddOuterScopeNodeArg for the initializer in the parent graph when creating the subgraph.

Reply:
Changes are in https://github.com/microsoft/onnxruntime/pull/1346, however I'm not convinced the initializer replacement code is valid (seems to assume an initializer is only ever used by one node) so it may need further refinement. 

I'm fine if you want to wait and use a separate PR to update this code and add a test after 1346 is completed.

Reply:
The initializer replacement code creates new names and NodeArgs for the reordered filters and aligned biases. These are stored in maps hanging off the impl class and are also used to only do the reordering once. There are also test cases (ConvReuseWeights*) to exercise these paths.

I first hit this issue with faster-rcnn which has three Conv nodes that reuse the same weights and my old logic was busted.

Did you have a more specific concern that I can deal with?

Reply:
Sorry - I meant the existing optimizer code touched in #1346 seemed to have issues not this PR. I'm working on fixing that.